src/main/java/com/exactpro/th2/mstore/ErrorCollector.java

```diff
@@ -0,0 +1,185 @@
+        this.drainFuture = requireNonNull(executor, "Executor can't be null")
+                .scheduleAtFixedRate(this::drain, period, period, unit);
+    }
+
+    public ErrorCollector(@NotNull ScheduledExecutorService executor, @NotNull MessageRouter<EventBatch> eventRouter,
+                          @NotNull EventID rootEvent) {
+        this(executor, eventRouter, rootEvent, 1, TimeUnit.MINUTES);
+    }
+
+    /**
+     * Log error and call the {@link #collect(String)}} method
+     * @param error is used as key indetifyer. Avoid put a lot of unique values
+     */
+    public void collect(Logger logger, String error, Throwable cause) {
+        logger.error(error, cause);
+        collect(error);
+    }
+
+    /**
+     * @param error is used as key indetifyer. Avoid put a lot of unique values
+     */
+    public void collect(String error) {
+        lock.lock();
+        try {
+            errors.compute(error, (key, metadata) -> {
+                if (metadata == null) {
+                    return new ErrorMetadata();
+                }
+                metadata.inc();
+                return metadata;
+            });
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        drainFuture.cancel(true);
+        drain();
+    }
+
+    private void drain() {
+        try {
+            Map<String, ErrorMetadata> map = clear();
+            if (map.isEmpty()) { return; }
+
+            eventRouter.sendAll(Event.start()
+                            .name("mstore internal problem(s): " + calulateTotalQty(map.values()))
+                            .type("InternalError")
+                            .status(Status.FAILED)
+                            .bodyData(new BodyData(map))
+                            .toBatchProto(rootEvent));
+
+        } catch (IOException | RuntimeException e) {
+            LOGGER.error("Drain events task failyre", e);
+        }
+    }
+
+    private Map<String, ErrorMetadata> clear() {
+        try {
+            lock.lock();
+            Map<String, ErrorMetadata> result = errors;
+            errors = new HashMap<>();
+            return result;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private static int calulateTotalQty(Collection<ErrorMetadata> errors) {
+        return errors.stream()
+                .map(ErrorMetadata::getQuantity)
+                .reduce(0, Integer::sum);
+    }
+
+    private static class BodyData implements IBodyData {
+        private final Map<String, ErrorMetadata> errors;
+        @JsonCreator
+        private BodyData(Map<String, ErrorMetadata> errors) {
+            this.errors = errors;
+        }
+        public Map<String, ErrorMetadata> getErrors() {
+            return errors;
+        }
+    }
+
+    private static class ErrorMetadata {
+        @JsonSerialize(using = InstantSerializer.class)
```

Comment:
Probably, we should register JavaTimeModule in ObjectMapper which is used to serialize body data in `Event` class. However,  I think it is okay to do it with explicit serializer for now

Reply:
I agree